alibi_detect/utils/pytorch/kernels.py

```diff
@@ -88,6 +168,271 @@ def forward(self, x: Union[np.ndarray, torch.Tensor], y: Union[np.ndarray, torch
+        self.feature_axis = feature_axis
+        self.trainable = trainable
+
+    @property
+    def alpha(self) -> torch.Tensor:
+        return self.raw_alpha
+
+    @property
+    def sigma(self) -> torch.Tensor:
+        return self.log_sigma.exp()
+
+    def forward(self, x: Union[np.ndarray, torch.Tensor], y: Union[np.ndarray, torch.Tensor],
+                infer_parameter: bool = False) -> torch.Tensor:
+        x, y = torch.as_tensor(x), torch.as_tensor(y)
+        if self.active_dims is not None:
+            x = torch.index_select(x, self.feature_axis, self.active_dims)
+            y = torch.index_select(y, self.feature_axis, self.active_dims)
+        dist = distance.squared_pairwise_distance(x.flatten(1), y.flatten(1))
+
+        if infer_parameter or self.init_required:
+            if self.trainable and infer_parameter:
+                raise ValueError("Gradients cannot be computed w.r.t. an inferred sigma value")
+            sigma = self.init_fn_sigma(x, y, dist)
+            alpha = self.init_fn_alpha(x, y, dist)
+            with torch.no_grad():
+                self.log_sigma.copy_(sigma.log().clone())
+                self.raw_alpha.copy_(alpha.clone())
+            self.init_required = False
+
+        if len(self.sigma) > 1:
+            if len(self.sigma) == len(self.alpha):
+                kernel_mat = []
+                for i in range(len(self.sigma)):
+                    kernel_mat.append((1 + torch.square(dist)
+                                       / (2 * self.alpha[i] * (self.sigma[i] ** 2))) ** (-self.alpha[i]))
+                kernel_mat = torch.stack(kernel_mat, dim=0).mean(dim=0)
+            else:
+                raise ValueError("Length of sigma and alpha must be equal")
+        else:
+            kernel_mat = (1 + torch.square(dist) / (2 * self.alpha * (self.sigma ** 2))) ** (-self.alpha)
+        return kernel_mat
+
+
+class Periodic(BaseKernel):
+    def __init__(
+        self,
+        tau: torch.Tensor = None,
+        init_fn_tau: Callable = pseudo_init_fn,
+        sigma: torch.Tensor = None,
+        init_fn_sigma: Callable = sigma_median,
+        trainable: bool = False,
+        active_dims: Optional[list] = None,
+        feature_axis: int = -1
+    ) -> None:
+        """
+        Periodic kernel: k(x,y) = .
+        A forward pass takesa batch of instances x [Nx, features] and y [Ny, features]
+        and returns the kernel matrix [Nx, Ny].
+
+        Parameters
+        ----------
+        tau
+            Period of the periodic kernel.
+        sigma
+            Bandwidth used for the kernel.
+        """
+        super().__init__()
+        self.parameter_dict['tau'] = 'period'
+        self.parameter_dict['sigma'] = 'bandwidth'
+        if tau is None:
+            self.log_tau = nn.Parameter(torch.empty(1), requires_grad=trainable)
+            self.init_required = True
+        else:
+            self.log_tau = nn.Parameter(tau.log(), requires_grad=trainable)
```

Comment:
If we are going to the effort to make parameters trainable we should probably make it possible to train one and not the other, particularly given the period will (almost always?) be fixed according to domain specific knowledge whereas one may wish to train the bandwidth.

Reply:
The new parameter implementation should allow such behaviour. But at the moment for existing kernels, these are still specified by the global `trainable` argument. Guess the better solution here is to add a `trainable` option for each parameter.